### PR TITLE
[Get] Only instantiate Repo objects once for items in `Packages`.

### PR DESCRIPTION
**Pull request by ddunbar**
- The old algorithm was O(N^2)... this gives a 3x speedup in null build time
   for a large project (Zewo).

### CODE REVIEW

**Comment by ddunbar** [comment]
@mxcl can you review at some point? I'm not entirely sure this is safe, but it is a big win if so.

**Comment by mxcl** [comment]
Possibly a problem because repo paths change during the Get and Update processes based on which checkouts are selected and the Git.Repo path is not changed.

Probably it's fine because the renames happen last. But may be better to just figure out why `Repo.init` is slow.

**Comment by ddunbar** [comment]
It is slow because it shells out to git to check things... that said I'm worried about anything which is O(N^2) even if we made it much cheaper...

I was worried about the situations you mentioned, but couldn't get it to fail in any cases. We could always force invalidation like the last bit of the patch does in any other places that updates the repo.

What should we do here? I'd really like to see something like this merged just for the Zewo perf win.

**Comment by mxcl** [comment]
The initializer is:

```swift
    public init?(path: String) {
            guard let realroot = try? realpath(path) else { return nil }
            self.path = realroot
            guard Path.join(path, ".git").isDirectory else { return nil }
        }
```

It seems like `origin` is the slower function?

However I imagine this is safe since you identified the place where invalidation should happen.

I worry about this sort of thing however. The slow part of all this is the HTTP.

File: Sources/Get/PackagesDirectory.swift [hunk 1]
```diff
@@ -25,15 +25,23 @@ class PackagesDirectory {
...
+            result[origin] = repo
+        }
+        return result
+    }()
 }
 
 extension PackagesDirectory: Fetcher {
     typealias T = Package
-
+
```
Comment (kostiakoval) [inline Sources/Get/PackagesDirectory.swift]: remove whitespace :)

**Comment by ddunbar** [comment]
I'm going to go ahead and take this once it passes CI, if there is fallout then that is probably useful to identify and codify in test cases, and I really want to see the perf win land.

**Comment by ddunbar** [comment]
@swift-ci please test and merge

**Comment by ddunbar** [comment]
Linux failure is unrelated.

**Pull request merged**